### PR TITLE
Blog App: validations and model specs

### PR DESCRIPTION
**Pull request by Eccb7**
In this PR, I have:

- Added the following validations:
  - For the User model:
    - Name must not be blank.
    - PostsCounter must be an integer greater than or equal to zero.
  - For the Post model:
    - Title must not be blank.
    - Title must not exceed 250 characters.
    - CommentsCounter must be an integer greater than or equal to zero.
    - LikesCounter must be an integer greater than or equal to zero.
- Added unit specs for all of your models' methods and validations.

### CODE REVIEW

File: spec/post_spec.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    user = User.create(name: 'Bob Smith', posts_counter: 1)
+    post = user.posts.new(title: 'Sample Post', comments_counter: -1, likes_counter: 0)
+    expect(post).to_not be_valid
+  end
+
+  it 'is valid with zero comments_counter' do
+    user = User.create(name: 'Eve Johnson', posts_counter: 4)
+    post = user.posts.new(title: 'Sample Post', comments_counter: 0, likes_counter: 2)
+    expect(post).to be_valid
+  end
```
Comment (KanzaTahreem) [inline spec/post_spec.rb]: - [ ] Good job for testing the `comments_counter` validation 👏However, as per [this project requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec.md#:~:text=Add%20unit%20specs%20for%20all%20of%20your%20models%27%20methods%20and%20validations), we need to test all the validations, therefore, kindly add unit tests for the validation `LikesCounter must be an integer greater than or equal to zero.`

File: spec/user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  it 'is not valid with a negative posts_counter' do
+    user = User.new(name: 'Jane Smith', posts_counter: -1)
+    expect(user).to_not be_valid
+  end
+
+  it 'is valid with zero posts_counter' do
+    user = User.new(name: 'Alice Johnson', posts_counter: 0)
+    expect(user).to be_valid
+  end
```
Comment (KanzaTahreem) [inline spec/user_spec.rb:22]: - [ ] Good job so far ✨ However, I noticed that the unit tests for `recent_posts` method present in the `User` model are missing ⚠Similarly the other models' methods are also missing the unit tests. Kindly note that as per [this project requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec.md#:~:text=Add%20unit%20specs%20for%20all%20of%20your%20models%27%20methods%20and%20validations.), we need to add unit specs for ALL(`User`, `Post`, `Comment`, and `Like`) of your models' methods ⛏💪


**_Kindly make sure to test ALL of your models' methods (Please apply this to all models)_**

image

**Review by KanzaTahreem — CHANGES_REQUESTED** [review]
# Changes Required 🔁

Hi @Eccb7 

Good job so far 👏! I really appreciate the effort you put in!  **_However, there are few this that still need to be fixed. But you are almost there!_** 💪 🏁

so close


## Highlights 🎯

There are some excellent points that you have made perfectly. Keep up the great work! 👏
- ✅ `User` & `Post` models have validations
- ✅ Descriptive PR and README file
- ✅ Linters are working properly
- ✅ All implemented tests are passing

image


## Required Changes 🛑

_Please check the comments under the review._

## Optional suggestions ❕

N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👨‍💻💻

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/post_spec.rb [hunk 3]
```diff
@@ -13,15 +19,16 @@
...
   end
 
-  it 'is valid with zero comments_counter' do
+  it 'is valid with zero likes_counter' do
     user = User.create(name: 'Eve Johnson', posts_counter: 4)
-    post = user.posts.new(title: 'Sample Post', comments_counter: 0, likes_counter: 2)
+    post = user.posts.create(title: 'Sample Post', comments_counter: 3, likes_counter: 0)
     expect(post).to be_valid
   end
 end
```
Comment (HAMMAS-SALEEM) [inline spec/post_spec.rb]: - [ ] Great work on making your test pass but i see that previous reviewer asked you to add some more tests for the likes counter but you have modified the comments_counter tests for likes_counter 😨 Now you don't have any test to verify the comments_counter 😮 Kindly make separate tests for the comments_counter and separate for likes_counter to follow the project requirements 👍

File: spec/user_spec.rb [hunk 4]
```diff
@@ -20,4 +20,14 @@
...
   end
+
+  it 'returns the most recent posts' do
+    user = User.create(name: 'John Doe', posts_counter: 5)
+    older_post = user.posts.create(title: 'Old Post', comments_counter: 2, likes_counter: 3, created_at: 3.days.ago)
+    newer_post = user.posts.create(title: 'New Post', comments_counter: 2, likes_counter: 3, created_at: 1.day.ago)
+
+    recent_posts = user.recent_posts
+
+    expect(recent_posts).to eq([newer_post, older_post])
```
Comment (HAMMAS-SALEEM) [inline spec/user_spec.rb:31]: - [ ] Great work on testing the User model methods but I guess you missed the requirement to test all the methods from all the models 🤔 Kindly add tests for the Comment and Like model methods to follow the project requirements as suggested by the previous reviewer 👍

**Important Note: Kindly make separate files to test each model 👍 such as comment_spec.rb and like_spec.rb**

File: spec/post_spec.rb [hunk 5]
```diff
@@ -13,15 +19,16 @@
     expect(post).to_not be_valid
   end
```
Comment (HAMMAS-SALEEM) [inline spec/post_spec.rb]: **[OPTIONAL BUT HIGHLY RECOMMENDED]**

- [ ] It would be nice if you make your own logo of your name and put that in the readme file at the top to present yourself as a brand in the job market or tech industry to catch the recruiters' eyes 👍

**Review by HAMMAS-SALEEM — CHANGES_REQUESTED** [review]
# Required Changes :recycle:
Ukhv

## Hey @Eccb7 :wave:

## Highlights :1st_place_medal:

✔️ Descriptive PR Title
✔️ Descriptive PR Summary
✔️ Following Gitflow
✔️ Following some requirements
✔️ Great work on testing the User model custom methods 🔥

Good job on implementing some of the requested changes, There are some issues that you still need to work on to go to the next project but you are almost there! :blush:

# Optional

_All the changes with optional tags are not crucial enough to prevent you from getting the approval but we highly recommend to students that they should take these changes into consideration._

Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

---------------------------------------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by KayLemba — APPROVED** [review]
[[# STATUS: APPROVED :green_circle: :clap: 

Hello @Eccb7    , 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  . 
Congratulations! 🎉  Keep up the Awesome work 🥂  Keep Rocking 🥇 

HIGHLIGHTS:

- [x]  Your repo is professional  😎 
- [x] Descriptive P.R 💯 
- [x] Professional README file  
- [x] Tests are passing 






 

 










 








Now you can merge this branch and move on :1st_place_medal: 
Happy Coding cheers 🥇 💯 



well-1

Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.



Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._]]

**Comment by Eccb7** [comment]
Thanks @KayLemba
